### PR TITLE
[#37] Add JSON schema for probes validation

### PR DESCRIPTION
**Pull request by yersan**
Follows up on #37

Adds the JSON schema for Probes validation

### CODE REVIEW

File: charts/wildfly/values.schema.json [hunk 1]
```diff
@@ -594,13 +594,499 @@
...
+                            },
+                            {
+                              "type": "integer"
+                            }
+                          ]
+                        }
+                      }
+                    },
+                    "terminationGracePeriodSeconds": {
+                      "description": "Optional duration in seconds the pod needs to terminate gracefully upon probe failure. The grace period is the duration in seconds after the processes running in the pod are sent a termination signal and the time when the processes are forcibly halted with a kill signal. Set this value longer than the expected cleanup time for your process. If this value is nil, the pod's terminationGracePeriodSeconds will be used. Otherwise, this value overrides the value provided by the pod spec. Value must be non-negative integer. The value zero indicates stop immediately via the kill signal (no opportunity to shut down). This is a beta field and requires enabling ProbeTerminationGracePeriod feature gate. Minimum value is 1. spec.terminationGracePeriodSeconds is used if unset.",
```
Comment (laDok8) [inline charts/wildfly/values.schema.json:744]: \> The value zero indicates stop immediately via the kill signal
\> Minimum value is 1
stopping immediately is not an option ?

Consider 
"If this value is nil OR UNSET, the pod's terminationGracePeriodSeconds will be used"
and removing "spec.terminationGracePeriodSeconds is used if unset."," to make desc shorter

File: charts/wildfly/values.schema.json [hunk 2]
```diff
@@ -594,13 +594,499 @@
...
+                      "format": "int64",
+                      "type": [
+                        "integer",
+                        "null"
+                      ]
+                    },
+                    "timeoutSeconds": {
+                      "description": "Number of seconds after which the probe times out. Defaults to 1 second. Minimum value is 1. More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#container-probes",
```
Comment (laDok8) [inline charts/wildfly/values.schema.json:752]: although optional minimum (and probably also maximum) can also be enforced in schema. see example here: https://helm.sh/docs/topics/charts/

File: charts/wildfly/values.schema.json [hunk 3]
```diff
@@ -594,13 +594,499 @@
...
+                      "properties": {
+                        "host": {
+                          "description": "Optional: Host name to connect to, defaults to the pod IP.",
+                          "type": "string"
+                        },
+                        "port": {
+                          "description": "Number or name of the port to access on the container. Number must be in the range 1 to 65535. Name must be an IANA_SVC_NAME.",
+                          "oneOf": [
+                            {
+                              "type": "string"
```
Comment (laDok8) [inline charts/wildfly/values.schema.json:734]: why is string allowed?

File: charts/wildfly/values.schema.json [hunk 4]
```diff
@@ -594,13 +594,499 @@
...
+                          "type": "string"
+                        }
+                      },
+                      "required": [
+                        "port"
+                      ],
+                      "type": "object"
+                    },
+                    "initialDelaySeconds": {
+                      "description": "Number of seconds after the container has started before liveness probes are initiated. More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#container-probes",
```
Comment (laDok8) [inline charts/wildfly/values.schema.json:861]: copy paste error `readiness probe` (more occurrences further down)

File: charts/wildfly/values.schema.json [hunk 5]
```diff
@@ -594,13 +594,499 @@
...
+                    "periodSeconds": {
+                      "description": "How often (in seconds) to perform the probe. Default to 10 seconds. Minimum value is 1.",
+                      "format": "int32",
+                      "type": [
+                        "integer",
+                        "null"
+                      ]
+                    },
+                    "successThreshold": {
+                      "description": "Minimum consecutive successes for the probe to be considered successful after having failed. Defaults to 1. Must be 1 for liveness and startup. Minimum value is 1.",
```
Comment (laDok8) [inline charts/wildfly/values.schema.json:877]: "Must be 1 for liveness and startup", not relevant here

File: charts/wildfly/values.schema.json [hunk 6]
```diff
@@ -594,13 +594,499 @@
                   }
                 },
                 "livenessProbe": {
-                    "description": "Freeform livenessProbe configuration"
+                  "type": "object",
+                  "description": "livenessProbe describes a health check to be performed against a container to determine whether it is alive.",
+                  "properties": {
+                    "exec": {
+                      "description": "Exec specifies the action to take.",
```
Comment (laDok8) [inline charts/wildfly/values.schema.json:601]:
```suggestion
                      "description": "Exec specifies actions to take.",
```

File: charts/wildfly/values.schema.json (hunk 1 above)
Comment (laDok8) [reply charts/wildfly/values.schema.json:744]: good idea

**Comment by jmesnil** [comment]
When I create the JSON schema for the Helm Chart, I let some elements unspecified on purpose.
All elements that are actually passed through the Helm Chart to the underlying Kubernetes cluster are "free form" so that we are not tied to any specific version of Kubernetes.
That is the case for the following stanza:

* `build.env`
* `build.images`
* `build.resources`
* `deploy.env`
* `deploy.envFrom`
* `deploy.extraContainers`
* `deploy.initContainers`
* `deploy.livenessProbe`
* `deploy.readinessProbe`
* `deploy.resources`
* `deploy.startupProbe`
* `deploy.volumes`
* `deploy.volumeMounts`

If we enforce their validation by adding them to the Chart's schema, we might have issues if the underlying resources have a different schema in future versions.
This can become a big maintenance burden to ensure that the Chart schema is "in sync" with Kubernetes (and which version of Kubernetes would be used? latest? a baseline?)

**Comment by laDok8** [comment]
valid concern. I  don't see problem with keeping current free form

**Comment by yersan** [comment]
@jmesnil @laDok8 I agree with that, I proceeded to elaborate on this schema as a result of our initial discussions, but I am more than happy if we now decide to leaving these fields as open form. At the end, the resource will be rejected by the Kubernetes API, so this would be more of an utility for third-party tools or to anticipate the creation of any resource if the schema is rejected. 


> When I create the JSON schema for the Helm Chart, I let some elements unspecified on purpose.

@jmesnil My only question about this is if Helm Charts can create partial resources if one of them is finally rejected by the Kubernetes API because it does not have the correct schema. It would be a weird behavior of Helm Charts though, but just to be sure if that can happened. I would expect Helm Charts creates everything at once and if one resource is rejected, all the previous resources successfully created before the rejected one are not "committed" to the, etc database.

Otherwise, I'm fine if we close this and leave the fields in a free form.

**Comment by yersan** [comment]
As per discussions, close this as won't do.

**Pull request closed without merging**